pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -175,3 +207,145 @@ func (t *Terminator) podDeleteTimeWithGracePeriod(nodeGracePeriodExpirationTime
 	deleteTime := nodeGracePeriodExpirationTime.Add(time.Duration(*pod.Spec.TerminationGracePeriodSeconds) * time.Second * -1)
 	return &deleteTime
 }
+
+func (t *Terminator) GetDeploymentFromPod(ctx context.Context, pod *corev1.Pod) (*appsv1.Deployment, error) {
+	rs, err := t.getOwnerReplicaSet(ctx, pod)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ReplicaSet from Pod: %w", err)
+	}
+	if rs == nil {
+		return nil, nil
+	}
+
+	deployment, err := t.getOwnerDeployment(ctx, rs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Deployment from ReplicaSet: %w", err)
+	}
+	return deployment, nil
+
+}
+
+func (t *Terminator) getOwnerReplicaSet(ctx context.Context, pod *corev1.Pod) (*appsv1.ReplicaSet, error) {
+	for _, ownerRef := range pod.GetOwnerReferences() {
+		if ownerRef.Controller != nil && ownerRef.Kind == "ReplicaSet" {
+			rs := &appsv1.ReplicaSet{}
+			if err := t.kubeClient.Get(ctx, client.ObjectKey{Name: ownerRef.Name, Namespace: pod.Namespace}, rs); err != nil {
+				return nil, fmt.Errorf("get ReplicaSet: %w", err)
+			}
+			return rs, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (t *Terminator) getOwnerDeployment(ctx context.Context, rs *appsv1.ReplicaSet) (*appsv1.Deployment, error) {
+	for _, ownerRef := range rs.GetOwnerReferences() {
+		if ownerRef.Controller != nil && ownerRef.Kind == "Deployment" {
+			deployment := &appsv1.Deployment{}
+			if err := t.kubeClient.Get(ctx, client.ObjectKey{Name: ownerRef.Name, Namespace: rs.Namespace}, deployment); err != nil {
+				return nil, fmt.Errorf("get Deployment: %w", err)
+			}
+			return deployment, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (t *Terminator) RestartDeployments(ctx context.Context, deployments []*appsv1.Deployment, nodeName string) error {
+	var updateErrors []error
+
+	for _, deployment := range deployments {
+		if deployment.Spec.Template.Annotations == nil {
+			deployment.Spec.Template.Annotations = make(map[string]string)
+		}
+		restartedNode, exists := deployment.Spec.Template.Annotations[v1.DeploymentRestartNodeAnnotationKey]
+		if exists && restartedNode == nodeName {
```

Comment:
Why skip if `restartedNode == nodeName`? And other annotation value is better, like `true`?

Reply:
Why not take the annotation from `deployment.Annotations`?

Reply:
Updating deployment.Spec.Template.Annotations can restart deployment, but updating deployment.Annotations will not cause deployment to restart

Here deployment.Spec.Template.Annotations is used as a sign of whether deployment has been restarted. If annotation already has nodeName, it means that the last Reconcile has restarted deployment, so it will be skipped this time

After testing, it can be implemented. Can you help to see if it is suitable?